src/block/Hopper.php

```diff
@@ -92,8 +99,231 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
 	}
 
 	public function onScheduledUpdate() : void{
-		//TODO
+		$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 1);
```

Comment:
If the block above is not a container, but a full block, there's no point in ticking the hopper all the time while the transfer is in cooldown for x amount of time if the block below is a container.  
In addition, if the hopper has a full block above it but no container in front of it, the tick should not be activated to improve performance.

Reply:
As said above, there still could be item entities in that full block.